curated_transformers/layers/transformer.py

```diff
@@ -3,12 +3,197 @@
 
 import torch
 from torch import Tensor
-from torch.nn import Dropout, Identity, Module
+from torch.nn import Dropout, Embedding, Identity, LayerNorm, Linear, Module
 
 from .attention import AttentionMask, KeyValueCache, SelfAttention
 from .feedforward import PointwiseFeedForward
 
 
+@dataclass
+class EmbeddingsLayerNorms:
+    """
+    Layer normalizations used in a transformer embedding layer.
+
+    By default, all the normalizations are disabled by setting the layer
+    normalization to the Torch ``Identity`` module. Therefore, only
+    normalizations that are needed have to be set.
+
+    :param embed_output_layer_norm:
+        Normalization of the embeddings.
+    :param proj_output_layer_norm:
+        Normalization of the output of the projection layer.
+    """
+
+    embed_output_layer_norm: Module = Identity()
+    proj_output_layer_norm: Module = Identity()
+
+
+@dataclass
+class EmbeddingsDropouts:
+    """
+    Dropouts used in a transformer embedding layer.
+
+    By default, all the dropouts are disabled by setting the dropout
+    to the Torch ``Identity`` module. Therefore, only dropouts that are
+    needed have to be set.
+
+    :param embed_output_dropout:
+        Dropout of the embeddings.
+    :param proj_output_dropout:
+        Dropout of the output of the projection layer.
+    """
+
+    embed_output_dropout: Module = Identity()
+    proj_output_dropout: Module = Identity()
+
+
+class TransformerEmbeddings(Module):
+    """
+    Transformer embeddings layer.
+
+    This is a generic transformer embedding layer. The layer alwas has piece
+    embeddings and can optionally have position embeddings, type embeddings,
+    and a projection of embeddings to the model's hidden size.
+    """
+
+    def __init__(
+        self,
+        *,
+        dropouts: EmbeddingsDropouts,
+        embedding_width: int,
+        hidden_width: int,
+        layer_norms: EmbeddingsLayerNorms,
+        n_pieces: int,
+        n_positions: Optional[int],
+        n_types: Optional[int],
```

Comment:
I think we use `num_` as the prefix for counts in other places.

Reply:
We should probably do a consistency sweep of things across the code base before the final release, but I think I'd prefer renaming all these `num_` cases to `n_`. I think `n` is a more widely used abbreviation for number and we only have `num_` because HF model configurations have this prefix?

Reply:
I'd personally prefer `n_` as well :-)

Reply:
Sounds good!